### DISCUSSION TITLE
Some questions about the hardware design of MobileNet

### DISCUSSION THREAD

**Discussion by sansi-zhang** [Xilinx/finn#897] [Q&A] [answered]
Hi @maltanar, I need to ask you a few questions.

## 1. Thoughts on "Creating a Dataflow Partition"
The process of "Creating a Dataflow Partition", as described in the notebook, is to separate out the HLS layer and handle it separately, while some of the usual onnx operations are retained.

I thought briefly, does it support normal operations other than 'Transpose'? If so, what can this so-called "regular operation" include?

### Note
I tried not to use the activation function and BN at the end of the node, I get an extra ```MUL``` node that works just fine when doing ```ZynqBuild```.

## 2. No longer need to specify fifo depth when designing MVAU

When designing the MVAU parameters, it is now only necessary to specify the way in which memory is used. But when I checked the folding_config of mobilenet in the "finn-examples" project, I found that it contains fifo depth information. Is it auto-configured? If not, how should it be configured?

## 3. The PE design basis of VVAU in MobileNet

I have checked the design process of mobilenet in "end2end". I do not quite understand the design basis of PE and SIMD of VVAU. Does it still need to satisfy "MW % SIMD == 0, MH % PE == 0"? Why it needs to design in the process "ConvInputGen" above the vvau SIMD?

And I found that my upper layer "FMPadding" node was of type FMPadding_rtl instead of "FMPadding_Batch" in the original process, so its SIMD did not change. I used this unchanged model to run subsequent tasks in turn, but when I ran to "ZynqBuild", Error: ```ValueError: cannot reshape array of size 675000 into shape (1,50,50,67,4)```. So I modified the type judgment, it changed the SIMD value, and I was able to run ```ZynqBuild``` just fine.

Thanks.

**Comment by fpjentzsch** [comment, marked as answer]
Hi, perhaps I can help.

1) In principle you can have any ONNX operation that is not supported by FINN in the graph. You can then synthesize the accelerator(s) for compatible dataflow partition(s) (containing only HLS layers) and handle the unsupported operations yourself by implementing them in software or manually inserting your own FPGA IP. Is this what you meant?

2) You can enable automatic FIFO sizing, see the builder config option(s) [here](https://github.com/Xilinx/finn/blob/13313c293eb00d2eb88edcaf68386fdf52152aeb/src/finn/builder/build_dataflow_config.py#L262C13-L262C13). For the MobileNet example, I believe we started with an automatic FIFO configuration and then optimized it by hand, resulting in the given folding_config. This is generally a good practice, as it saves build time for subsequent iterations (assuming folding is not changed) because the automatic sizing algorithm doesn't need to be run anymore.

3) Please see this paragraph in the documentation about VVAU folding: https://finn-dev.readthedocs.io/en/latest/internals.html#folding. What you observed with the folding config node names is expected, since the config checks the names very strictly and does not understand how the "FMPadding_rtl" (new RTL component) and "FMPadding_Batch" (old HLS component) variants might be related. We might rework this behavior in the future.

**Comment by sansi-zhang** [comment]
> Hi, perhaps I can help.
> 
> 
> 
> 1) In principle you can have any ONNX operation that is not supported by FINN in the graph. You can then synthesize the accelerator(s) for compatible dataflow partition(s) (containing only HLS layers) and handle the unsupported operations yourself by implementing them in software or manually inserting your own FPGA IP. Is this what you meant?
> 
> 
> 
> 2) You can enable automatic FIFO sizing, see the builder config option(s) [here](https://github.com/Xilinx/finn/blob/13313c293eb00d2eb88edcaf68386fdf52152aeb/src/finn/builder/build_dataflow_config.py#L262C13-L262C13). For the MobileNet example, I believe we started with an automatic FIFO configuration and then optimized it by hand, resulting in the given folding_config. This is generally a good practice, as it saves build time for subsequent iterations (assuming folding is not changed) because the automatic sizing algorithm doesn't need to be run anymore.
> 
> 
> 
> 3) Please see this paragraph in the documentation about VVAU folding: https://finn-dev.readthedocs.io/en/latest/internals.html#folding. What you observed with the folding config node names is expected, since the config checks the names very strictly and does not understand how the "FMPadding_rtl" (new RTL component) and "FMPadding_Batch" (old HLS component) variants might be related. We might rework this behavior in the future.


Thank you very much for your reply. 

For 1, my main meaning is as you described it. 
However, after I included some unsupported operations in the model as part of the onnx file, an error occurred when performing finn's operations. 
So I made the following adjustment: the unsupported parts were implemented by performing software operations in pynq.
